### DISCUSSION TITLE
Add AKWF Wavefrom

### DISCUSSION THREAD

**Discussion by GillesLACAUD** [marcel-licence/esp32_basic_synth#24] [Ideas]
You can try to add some of these waveforms :
https://www.adventurekid.se/akrt/waveforms/adventure-kid-waveforms/

Just apply a time convertion  with audacity for exemple to set the good sample number

it is very interresting for the voices waveform

**Comment by marcel-licence** [comment]
Merci. I can imagine that conversion can be scripted or maybe done internally in the firmware on the esp32. It would be handy just to upload the waveforms into the external flash memory and make it somehow available.

Good idea, best regards.

**Reply by ErichHeinemann** [reply]
Hi Marcel, I had already send the Link to the ".h" - files for these Waveforms :) You do not need to convert them by Yourself, someone other did it already https://github.com/DatanoiseTV/AKWF_WaveForms_1024

**Reply by GillesLACAUD** [reply]
Good information Erich. Thx for your help.
At my point i think a better way is to place all these samples in a file on the SD card and keep the offset for each individual waveform. So you open the file at the beginning and peek the 1024 samples at the good  offset.
I have done this on my HANSY1010 mono synth.
What do you think ?

**Reply by ErichHeinemann** [reply]
Hi Gilles, as You are creating a lot of Synths and sell them, - have You tried the Sampler-Version from Marcel? I believe, You could simply use this as the Oscillator and then put all the AKWF-Files as WAV onto the SD-Card. I have seen the Ai_Thinker-Audio-Board on Your Website, - it seems You are well prepared and could test it. Which GPIO-Pins Do You use in Your hardware to get a second I2C up and running on the AI_Thinker? I am using GPIO5 for SDA and GPIO23 for SCL .. on Core 0 works with 50.000 Hz. :)

**Comment by ErichHeinemann** [comment]
Hi Gilles,
I asked for Your experience and preferences to Use I2C on the AI Thinker Audio Board.
If Marcel starts to use the SD-Card, he probably will loose the free Pins to connect a ADC for local Potentiometers.

If he uses the SD-Card, a working and simple solution to create a new ADC-Port would be needed.

At the current stage, the Synth could be rebuild by many people. As You are in that business, You know, there is probably a hard limit in complexity for DIY-Projects. If the project gets too complex, only a few people could follow Marcel and could build these synths.

Regards
Erich

**Reply by GillesLACAUD** [reply]
Yes of course you are totaly right. For my part my first goal is to keep the Midi interface to control the parameter and use the touch screen for some other parameter (it is a Nextion screen). If i have to make a more complex IHM with encoder or potentiometer i will add another CPU for this with many more GPIO like an STM32L432 Nucleo Board. The ESP32 A1S is a very good processor but with not enough pin.
A little demo of my project
https://www.youtube.com/watch?v=2i6vPzySrOI

**Comment by ErichHeinemann** [comment]
Hi, I have tested it, but it didn´t sound that much.

How to modify the code to test it:
1. create a new file int he Arduino-Project called AKWF.h
2. Copy a Codeblock / Waveform  from https://github.com/DatanoiseTV/AKWF_WaveForms_1024 with one array into the new file.
3. Change the definition of the Array a bit, replace the sign "=" with PROGMEM to get this:
const int16_t AKWF_dbass_0001[1024] PROGMEM {
   347,   560, 
4. Change the count from 1025 to 1024 and delete the first or last value.
5. Save the file
6. got to file easysynth.ino and add this line at the beginning:
#include "AKWF.h"
7. Now we need to get the Wavform accessable as Array of Floats..
We change simply "silence" for testing, search for silence[i] = 0; and change it to:
        silence[i] = (float)pgm_read_word_near( AKWF_dbass_0001 + i )/32000.0;

8. Save and Compile and Upload it to the ESP32
When the ESP32 starts, change the Wavform to "6" - Silence. You could monitor it using the Serial Monitor in the IDE.

It is up to You to search for a cool sounding Waveform ...  
You could even let the arraydefinition stay at 1025 but it seems to be something like a Lego-brick with the size of 4,0001  - a bad size. 
If You add more than one Array into progmem, You could probably create a function to load the array on command into the wavform "silence". 
This was a quick test, done in 10 minutes. If someone finds some good sounding examples - let me know. For the moment, the build in  basic waveforms sound better for me.

**Reply by GillesLACAUD** [reply]
i will send you a demo with a voice waveform....

**Reply by MichaelPNolan** [reply]
Forgive me for being a bit new/lost and trying to clarify what seems obvious. Merci for your time.

@GillesLACAUD 
I'm interested to make a library like this and play -  any chance you can make a module or code example. I can see above you made a guide and I'm going to give it a go. Its only going to read the first WAVEFORM_CNT samples and then expand it to float with /32000.0 being like 16 bit to fit -1 to 1 range? Yes?

So I got what looks like a 16bit wav sample array for a violin. I'm guessing there is one sample for each octave? So I took something in the middle. So yes, it works on my little synth and sounds - ok I guess ... agree would be good to share a some good samples people tried. I will if I find any. Also interested in mathematically generating some. Perhaps some timbre with a 5th overtone

My project has a fair bit of new code but its still only 61% of program storage and 14% of dynamic memory - so can I put the sample in dynamic memory? Will it just go there if I take away PROGMEM - any reason you put in PROGMEM - is preserving dynamic memory important here - I'm thinking out loud. I guess yes ... when all this gets loaded.

Marcel's waves seem only to be 40 bytes long ... so I feel like I completely don't understand 
@marcel-licence 
Even though I kind of understand - I wouldn't mind to get your comments, if you are willing, on how I understand the size or a link to education in case I just am not intelligent enough

```
#define WAVEFORM_BIT	10UL   //U unsigned L long so 40 bytes? 
#define WAVEFORM_CNT	(1<<WAVEFORM_BIT)
#define WAVEFORM_Q4		(1<<(WAVEFORM_BIT-2))
#define WAVEFORM_MSK	((1<<WAVEFORM_BIT)-1)
#define WAVEFORM_I(i)	((i) >> (32 - WAVEFORM_BIT)) & WAVEFORM_MSK
```

Because I want to either make some more formulas or copy an array of data in.

My stand-alone keyboard is coming along well and I've started making arpeggiator and probably write a simple sequencer and then make a video of myself jamming. I wouldn't mind adding a few waveforms for that ... I'm trying to make something relatively fully featured to entice others who might want to do something with your code or my fork.

**Reply by GillesLACAUD** [reply]
it is /32768 to fit -1 to 1 
The waveform is 1024 long not 40 it is a logical 1 shift left 10 times (WAVEFORM_BIT=10) so 2*2*2*2*2*2*2*2*2*2=1024.
On my code all the AKWF waveform (transfrom to 1024 samples) are store on the SD card. So i just need 1024 float in the memory.
Please see my github repo for more information.
I will add some other docs as soon as possible

Some sounds demo from my project
https://www.youtube.com/watch?v=Dmk6--IBhCQ

**Reply by MichaelPNolan** [reply]
Thanks - its 10 times - now I see it ... I'm not used to reading those bit operations and I forget. So this helps and as I said I got the sample to work with relative ease thanks to your advice. So .. thats great.
Now I know why it's called WAVEFORM_BIT - its the BIT SHIFTING index ... 2^10 via binary.

**Comment by GillesLACAUD** [comment]
The demo for the voice wave form it is at the end of the vid

https://www.youtube.com/watch?v=gsOaK0aHd1I

**Reply by MichaelPNolan** [reply]
Thats a great thing ... I gotta demo my thing in here. Its pretty kudgy but it starts to do some good things. You have a whole ton of work its great to see. I hope I can see some of your code and add functions to my projects. I'm doing this a lot of my spare time.

**Comment by ErichHeinemann** [comment]
Hi Gilles,

which AKWF-Files did You use?
As I pushed the code on my repo today.
Was it ok?

I did test with some simple AKWF-Files but did not get very special sounds.
Which did You use?

Do You use them from Program?

As You do not allow discussion on Youtube and You do not have any results at Github, it is a bit difficult to talk about Your work.

When using GPL-Software, I allways try to give it back to the community.How much of the ESP-Synth is based on the Basic-Synth from Marcel?
As the Oscillators You are using in the Video  are the same as Marcel used ??

**Comment by GillesLACAUD** [comment]
Erich.

If you choice some simple AKWF-Files you will not have some special sounds.
You should try a voice waveform or an Epiano.
You can choice them with the help of this link.
https://www.adventurekid.se/AKRTfiles/AKWF/view/waveforms_index.html

For now i have code the voice waveform in the .h file as you (not read from the SD card). But me i have rebuild the tabwavefrom with the help of Audacity and Hxd.

For the GPL code...i try to give to you some ideas and information about my work. My code is now very different as the first commit of Marcel. I use another IDE (plateform IO) and some other hardware part...

I will take a look at your code...

**Comment by GillesLACAUD** [comment]
Why 35000 ? 
it is 32768 no ?

silence[i] = (float) pgm_read_word_near( AKWF_hvoice_0001 + i )/35000.0;

**Reply by marcel-licence** [reply]
Maybe defines from limits.h are useful: https://www.tutorialspoint.com/c_standard_library/limits_h.htm

**Comment by GillesLACAUD** [comment]
Hi Erich 


My project based on the Marcel licence work is now available on github 


GITHUB LINK 



Best 
Musically 
Gilles/Hansy Synth 




----- Mail original -----

De: "Erich" ***@***.***> 
À: "marcel-licence/esp32_basic_synth" ***@***.***> 
Cc: "Gilles Lacaud" ***@***.***>, "Author" ***@***.***> 
Envoyé: Vendredi 28 Mai 2021 03:32:28 
Objet: Re: [marcel-licence/esp32_basic_synth] Add AKWF Wavefrom (#24) 


Hi Gilles, 

which AKWF-Files did You use? 
As I pushed the code on my repo today. 
Was it ok? 

I did test with some simple AKWF-Files but did not get very special sounds. 
Which did You use? 

Do You use them from Program? 

As You do not allow discussion on Youtube and You do not have any results at Github, it is a bit difficult to talk about Your work. 

When using GPL-Software, I allways try to give it back to the community.How much of the ESP-Synth is based on the Basic-Synth from Marcel? 
As the Oscillators You are using in the Video are the same as Marcel used ??

**Comment by GillesLACAUD** [comment]
Hi Marcel 


I have buy some other ESP32 Audio Kit boards and of course now i have the ES8388 issue. 
So i have change the I2C pin to 18 and 23 and change the I2C address. 
In fact i use your ES8388 code. 
But my init ES8388 return always fail. 


Did i miss something ? 
There is an hardware modification ? 


Thank again for you help 


Musically 
Gilles/Hansy Synth
